[BUG] SIGSEGV when Prometheus instance becomes unavailable

```
❯ pint version                                                             
0.44.1 (revision: Homebrew)
```

`.pint.hcl`:

```hcl
rule {
  match {
    kind = "alerting"
  }

  label "severity" {
      value    = "(none|info|low|moderate|high|critical)"
      required = true
    }

}

prometheus "syn-p001" {
  uri     = "https://prometheus.xxx.yyy.com"
  timeout = "1m"
}
```

When running `pint lint` against some dummy rule:

```yaml
# pint ignore/next-line
namespace: test_alert
groups:
    - name: test_alert.rules
      rules:
        - alert: test_alert
          expr: absent(test_metric) == 1
          labels:
            severity: moderate
            team: team_devops
          annotations:
            description: ignore - test alert
            summary: ignore - test alert
```

Prometheus got OOM and pint paniced:

```text
❯ pint lint  infrastructure/base/mimir/rules/synerise/team_devops/test.yaml         
level=info msg="Loading configuration file" path=.pint.hcl
level=error msg="Query returned an error" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" query=count(up) uri=https://prometheus.xxx.yyy.com
level=error msg="Query returned an error" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" query=count(up) uri=https://prometheus.xxx.yyy.com
level=error msg="Query returned an error" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" query=count(up) uri=https://prometheus.xxx.yyy.com
level=error msg="Query returned an error" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" query=count(up) uri=https://prometheus.xxx.yyy.com
level=error msg="Query returned an error" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" query=count(up) uri=https://prometheus.xxx.yyy.com
level=error msg="Query returned an error" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" query=count(up) uri=https://prometheus.xxx.yyy.com
level=error msg="Query returned an error" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" query=count(up) uri=https://prometheus.xxx.yyy.com
level=error msg="Query returned an error" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" query=count(up) uri=https://prometheus.xxx.yyy.com
level=error msg="Query returned an error" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" query=count(up) uri=https://prometheus.xxx.yyy.com
level=error msg="Query returned an error" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" query=count(up) uri=https://prometheus.xxx.yyy.com
level=error msg="Query returned an error" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" query=count(up) uri=https://prometheus.xxx.yyy.com
level=error msg="Query returned an error" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" query=count(up) uri=https://prometheus.xxx.yyy.com
level=error msg="Query returned an error" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" query=count(up) uri=https://prometheus.xxx.yyy.com
level=error msg="Query returned an error" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" query=count(up) uri=https://prometheus.xxx.yyy.com
level=error msg="Query returned an error" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" query=count(up) uri=https://prometheus.xxx.yyy.com
level=error msg="Query returned an error" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" query=count(up) uri=https://prometheus.xxx.yyy.com
level=warn msg="Cannot detect Prometheus uptime gaps" error="Post \"https://prometheus.xxx.yyy.com/api/v1/query_range\": read tcp 10.0.101.174:54930->172.26.136.253:443: read: connection reset by peer" name=syn-p001
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x100f580b8]

goroutine 72 [running]:
github.com/cloudflare/pint/internal/checks.SeriesCheck.Check({0x140002fec80?}, {0x1017488f8, 0x14000310270}, {0x14000041340?, 0x3e?}, {0x140001802d0, 0x0, {{0x0, 0x0}, {0x0, ...}, ...}}, ...)
        github.com/cloudflare/pint/internal/checks/promql_series.go:208 +0xe08
main.scanWorker({0x1017488f8, 0x14000310270}, 0x1400012eb40?, 0x1400012eba0?)
        github.com/cloudflare/pint/cmd/pint/scan.go:210 +0x2d8
main.checkRules.func2()
        github.com/cloudflare/pint/cmd/pint/scan.go:80 +0x60
created by main.checkRules
        github.com/cloudflare/pint/cmd/pint/scan.go:78 +0x2ac
```

Maybe it is worth to add, that this Prometheus works as a read-only server, as we use Grafana Mimir to store metrics.

--- a/cmd/pint/tests/0080_lint_online.txt
+++ b/cmd/pint/tests/0080_lint_online.txt
@@ -12,6 +12,7 @@ cmp stderr stderr.txt
 -- stderr.txt --
 level=info msg="Loading configuration file" path=.pint.hcl
 level=warn msg="No results for Prometheus uptime metric, you might have set uptime config option to a missing metric, please check your config" metric=prometheus_ready name=prom1
+level=warn msg="Using dummy Prometheus uptime metric results with no gaps" metric=prometheus_ready name=prom1
 rules/1.yml:2 Warning: http_errors_total[2d] selector is trying to query Prometheus for 2d worth of metrics, but prometheus "prom1" at http://127.0.0.1:7080 is configured to only keep 1d of metrics history (promql/range_query)
  2 |   expr: rate(http_errors_total[2d]) > 0
 

--- a/docs/changelog.md
+++ b/docs/changelog.md
@@ -1,5 +1,12 @@
 # Changelog
 
+## v0.44.2
+
+### Fixed
+
+- Fixed a crash in `promql/series` check when Prometheus instance becomes
+  unavailable - #682.
+
 ## v0.44.1
 
 ### Fixed

--- a/internal/checks/promql_series.go
+++ b/internal/checks/promql_series.go
@@ -195,6 +195,27 @@ func (c SeriesCheck) Check(ctx context.Context, _ string, rule parser.Rule, entr
 				Str("metric", c.prom.UptimeMetric()).
 				Msg("No results for Prometheus uptime metric, you might have set uptime config option to a missing metric, please check your config")
 		}
+		if promUptime == nil || promUptime.Series.Ranges.Len() == 0 {
+			log.Warn().
+				Str("name", c.prom.Name()).
+				Str("metric", c.prom.UptimeMetric()).
+				Msg("Using dummy Prometheus uptime metric results with no gaps")
+			promUptime = &promapi.RangeQueryResult{
+				Series: promapi.SeriesTimeRanges{
+					From:  params.Start(),
+					Until: params.End(),
+					Step:  params.Step(),
+					Ranges: promapi.MetricTimeRanges{
+						{
+							Fingerprint: 0,
+							Labels:      labels.Labels{},
+							Start:       params.Start(),
+							End:         params.End(),
+						},
+					},
+				},
+			}
+		}
 
 		bareSelector := stripLabels(selector)
 

--- a/internal/checks/promql_series_test.go
+++ b/internal/checks/promql_series_test.go
@@ -1061,6 +1061,131 @@ func TestSeriesCheck(t *testing.T) {
 			},
 		},
 		{
+			description: "#3 metric present once with labels, failed baseline query",
+			content:     "- record: foo\n  expr: sum(found{job=\"abc\", cluster=\"dev\"})\n",
+			checker:     newSeriesCheck,
+			prometheus:  newSimpleProm,
+			problems: func(uri string) []checks.Problem {
+				return []checks.Problem{
+					{
+						Fragment: `found`,
+						Lines:    []int{2},
+						Reporter: checks.SeriesCheckName,
+						Text:     seriesSometimesText("prom", uri, `found`, "1w", "1d5m"),
+						Severity: checks.Warning,
+					},
+				}
+			},
+			mocks: []*prometheusMock{
+				{
+					conds: []requestCondition{
+						requireQueryPath,
+						formCond{key: "query", value: `count(found{cluster="dev",job="abc"})`},
+					},
+					resp: respondWithEmptyVector(),
+				},
+				{
+					conds: []requestCondition{
+						requireRangeQueryPath,
+						formCond{key: "query", value: `count(found)`},
+					},
+					resp: matrixResponse{
+						samples: []*model.SampleStream{
+							generateSampleStream(
+								map[string]string{},
+								time.Now().Add(time.Hour*24*-5),
+								time.Now().Add(time.Hour*24*-4),
+								time.Minute*5,
+							),
+						},
+					},
+				},
+				{
+					conds: []requestCondition{
+						requireRangeQueryPath,
+						formCond{key: "query", value: `count(found{job="abc"})`},
+					},
+					resp: matrixResponse{
+						samples: []*model.SampleStream{
+							generateSampleStream(
+								map[string]string{},
+								time.Now().Add(time.Hour*24*-5),
+								time.Now().Add(time.Hour*24*-5),
+								time.Minute*5,
+							),
+						},
+					},
+				},
+				{
+					conds: []requestCondition{
+						requireRangeQueryPath,
+						formCond{key: "query", value: `count(found{cluster="dev"})`},
+					},
+					resp: matrixResponse{
+						samples: []*model.SampleStream{
+							generateSampleStream(
+								map[string]string{},
+								time.Now().Add(time.Hour*24*-5),
+								time.Now().Add(time.Hour*24*-5).Add(time.Minute*10),
+								time.Minute*5,
+							),
+						},
+					},
+				},
+				{
+					conds: []requestCondition{
+						requireRangeQueryPath,
+						formCond{key: "query", value: `absent(found{job=~".+"})`},
+					},
+					resp: matrixResponse{
+						samples: []*model.SampleStream{
+							generateSampleStream(
+								map[string]string{},
+								time.Now().Add(time.Hour*24*-7),
+								time.Now().Add(time.Hour*24*-5).Add(time.Minute*-10),
+								time.Minute*5,
+							),
+							generateSampleStream(
+								map[string]string{},
+								time.Now().Add(time.Hour*24*-5).Add(time.Minute*5),
+								time.Now(),
+								time.Minute*5,
+							),
+						},
+					},
+				},
+				{
+					conds: []requestCondition{
+						requireRangeQueryPath,
+						formCond{key: "query", value: `absent(found{cluster=~".+"})`},
+					},
+					resp: matrixResponse{
+						samples: []*model.SampleStream{
+							generateSampleStream(
+								map[string]string{},
+								time.Now().Add(time.Hour*24*-7),
+								time.Now().Add(time.Hour*24*-5).Add(time.Minute*-10),
+								time.Minute*5,
+							),
+							generateSampleStream(
+								map[string]string{},
+								time.Now().Add(time.Hour*24*-5).Add(time.Minute*10),
+								time.Now(),
+								time.Minute*5,
+							),
+						},
+					},
+				},
+				{
+					conds: []requestCondition{
+						requireRangeQueryPath,
+						formCond{key: "query", value: "count(up)"},
+					},
+					resp: respondWithInternalError(),
+				},
+			},
+		},
+		{
 			description: "#4 metric was present but disappeared 50m ago",
 			content:     "- record: foo\n  expr: sum(found{job=\"foo\", instance=\"bar\"})\n",
 			checker:     newSeriesCheck,
@@ -3021,6 +3146,46 @@ func TestSeriesCheck(t *testing.T) {
 						formCond{key: "query", value: `count({__name__=~"(foo|bar)_panics_total",job="myjob"})`},
 					},
 					resp: respondWithEmptyMatrix(),
+				},
+			},
+		},
+		{
+			description: "series missing, failed baseline query",
+			content:     "- record: foo\n  expr: count(notfound) == 0\n",
+			checker:     newSeriesCheck,
+			prometheus:  newSimpleProm,
+			problems: func(uri string) []checks.Problem {
+				return []checks.Problem{
+					{
+						Fragment: `notfound`,
+						Lines:    []int{2},
+						Reporter: checks.SeriesCheckName,
+						Text:     noSeriesText("prom", uri, "notfound", "1w"),
+						Severity: checks.Bug,
+					},
+				}
+			},
+			mocks: []*prometheusMock{
+				{
+					conds: []requestCondition{
+						requireQueryPath,
+						formCond{key: "query", value: `count(notfound)`},
+					},
+					resp: respondWithEmptyVector(),
+				},
+				{
+					conds: []requestCondition{
+						requireRangeQueryPath,
+						formCond{key: "query", value: `count(notfound)`},
+					},
+					resp: respondWithEmptyMatrix(),
+				},
+				{
+					conds: []requestCondition{
+						requireRangeQueryPath,
+						formCond{key: "query", value: "count(up)"},
+					},
+					resp: respondWithInternalError(),
 				},
 			},
 		},

